File: src/main/java/pw/switchcraft/plethora/integration/vanilla/converter/VanillaConverters.java
package pw.switchcraft.plethora.integration.vanilla.converter;

import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import pw.switchcraft.plethora.api.WorldLocation;
import pw.switchcraft.plethora.api.converter.ConstantConverter;
import pw.switchcraft.plethora.api.converter.DynamicConverter;
import pw.switchcraft.plethora.api.reference.BlockReference;
import pw.switchcraft.plethora.integration.EntityIdentifier;

public final class VanillaConverters {
    public static final ConstantConverter<ItemStack, Item> GET_STACK_ITEM = ItemStack::getItem;

    public static final DynamicConverter<BlockReference, BlockState> GET_BLOCK_REFERENCE_BLOCK = BlockReference::getState;

    public static final ConstantConverter<BlockReference, BlockEntity> GET_BLOCK_REFERENCE_BLOCK_ENTITY = BlockReference::getBlockEntity;

    public static final ConstantConverter<BlockEntity, BlockReference> GET_BLOCK_ENTITY_REFERENCE = from -> {
        World world = from.getWorld();
        BlockPos pos = from.getPos();

        // Ensure we're referencing a valid TE
        return world == null || pos == null || world.getBlockEntity(pos) != from
            ? null
            : new BlockReference(new WorldLocation(world, pos), world.getBlockState(pos), from);
    };

    public static final ConstantConverter<BlockState, Block> GET_BLOCK_STATE_BLOCK = BlockState::getBlock;

    public static final ConstantConverter<Entity, EntityIdentifier> GET_ENTITY_IDENTIFIER = from ->
        from instanceof PlayerEntity player
            ? new EntityIdentifier.Player(player.getGameProfile())
            : new EntityIdentifier(from);
}

File: src/main/java/io/sc3/plethora/integration/vanilla/converter/VanillaConverters.kt
package io.sc3.plethora.integration.vanilla.converter

import io.sc3.plethora.api.WorldLocation
import io.sc3.plethora.api.converter.ConstantConverter
import io.sc3.plethora.api.converter.DynamicConverter
import io.sc3.plethora.api.reference.BlockReference
import io.sc3.plethora.integration.EntityIdentifier
import net.minecraft.block.BlockState
import net.minecraft.block.entity.BlockEntity
import net.minecraft.entity.Entity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemStack

object VanillaConverters {
  val GET_STACK_ITEM = ConstantConverter { obj: ItemStack -> obj.item }

  val GET_BLOCK_REFERENCE_BLOCK = DynamicConverter { obj: BlockReference -> obj.state }
  val GET_BLOCK_REFERENCE_BLOCK_ENTITY = ConstantConverter { obj: BlockReference -> obj.blockEntity }
  val GET_BLOCK_ENTITY_REFERENCE = ConstantConverter { from: BlockEntity ->
    val world = from.world
    val pos = from.pos
    if (world != null && pos != null && world.getBlockEntity(pos) === from) {
      BlockReference(WorldLocation(world, pos), world.getBlockState(pos), from)
    } else {
      null
    }
  }

  val GET_BLOCK_STATE_BLOCK = ConstantConverter { obj: BlockState -> obj.block }

  val GET_ENTITY_IDENTIFIER = ConstantConverter { from: Entity ->
    if (from is PlayerEntity) {
      EntityIdentifier.Player(from.gameProfile)
    } else {
      EntityIdentifier(from)
    }
  }
}
